src/main/java/com/android/volley/NetworkResponse.java

```diff
@@ -17,56 +17,136 @@
+            boolean notModified, long networkTimeMs) {
+        this(statusCode, data, toHeaderMap(allHeaders), allHeaders, notModified, networkTimeMs);
     }
 
+    /**
+     * Creates a new network response.
+     * @param statusCode the HTTP status code
+     * @param data Response body
+     * @param headers Headers returned with this response, or null for none
+     * @param notModified True if the server returned a 304 and the data was already in cache
+     * @deprecated see {@link #NetworkResponse(int, byte[], List, boolean, long)}. This constructor
+     *             cannot handle server responses containing multiple headers with the same name.
+     *             This constructor may be removed in a future release of Volley.
+     */
+    @Deprecated
     public NetworkResponse(int statusCode, byte[] data, Map<String, String> headers,
             boolean notModified) {
         this(statusCode, data, headers, notModified, 0);
     }
 
+    /**
+     * Creates a new network response for an OK response with no headers.
+     * @param data Response body
+     */
     public NetworkResponse(byte[] data) {
-        this(HttpURLConnection.HTTP_OK, data, Collections.<String, String>emptyMap(), false, 0);
+        this(HttpURLConnection.HTTP_OK, data, Collections.<Header>emptyList(), false, 0);
     }
 
+    /**
+     * Creates a new network response for an OK response.
+     * @param data Response body
+     * @param headers Headers returned with this response, or null for none
+     * @deprecated see {@link #NetworkResponse(int, byte[], List, boolean, long)}. This constructor
+     *             cannot handle server responses containing multiple headers with the same name.
+     *             This constructor may be removed in a future release of Volley.
+     */
+    @Deprecated
     public NetworkResponse(byte[] data, Map<String, String> headers) {
         this(HttpURLConnection.HTTP_OK, data, headers, false, 0);
     }
 
+    private NetworkResponse(int statusCode, byte[] data, Map<String, String> headers,
+            List<Header> allHeaders, boolean notModified, long networkTimeMs) {
+        this.statusCode = statusCode;
+        this.data = data;
+        this.headers = headers;
+        this.allHeaders = allHeaders;
+        this.notModified = notModified;
+        this.networkTimeMs = networkTimeMs;
+    }
+
     /** The HTTP status code. */
     public final int statusCode;
 
     /** Raw data from this response. */
     public final byte[] data;
 
-    /** Response headers. */
+    /**
+     * Immutable response headers.
+     *
+     * <p>This map is case-insensitive.
+     *
+     * <p>Note that if the server returns two headers with the same (case-insensitive) name, this
+     * map will only contain the last one. Use {@link #allHeaders} to inspect all headers returned
+     * by the server.
+     */
     public final Map<String, String> headers;
 
+    /** All response headers. */
```

Comment:
Why is the first header Map documented as immutable and not this one as well? (And are either of these really immutable?)

Reply:
Good question.

I've added a new commit to try and resolve this lack of clarity. The commit description there has full details, but in short - I've made all of the new List<Header> fields strictly immutable and documented them as such (with the exception of Cache.Entry, which is unenforceable because the field itself is mutable), and I've documented the legacy Map<String, String> fields as "should not be mutated" while not enforcing this to avoid breaking legacy apps.